component/http/v2/router/httprouter/router.go

```diff
@@ -0,0 +1,135 @@
+package httprouter
+
+import (
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/beatlabs/patron/component/http/middleware"
+	"github.com/beatlabs/patron/component/http/v2"
+	"github.com/beatlabs/patron/log"
+	"github.com/julienschmidt/httprouter"
+)
+
+const defaultDeflateLevel = 6
+
+// OptionFunc definition to allow functional configuration of the router.
+type OptionFunc func(*Config) error
+
+// Config definition.
+type Config struct {
+	aliveCheckFunc v2.LivenessCheckFunc
+	readyCheckFunc v2.ReadyCheckFunc
+	deflateLevel   int
+	middlewares    []middleware.Func
+	routes         []*v2.Route
+}
+
+func New(oo ...OptionFunc) (*httprouter.Router, error) {
+	cfg := &Config{
+		aliveCheckFunc: func() v2.AliveStatus { return v2.Alive },
+		readyCheckFunc: func() v2.ReadyStatus { return v2.Ready },
+		deflateLevel:   defaultDeflateLevel,
+	}
+
+	for _, option := range oo {
+		err := option(cfg)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var stdRoutes []*v2.Route
+
+	mux := httprouter.New()
+	stdRoutes = append(stdRoutes, v2.MetricRoute())
+	stdRoutes = append(stdRoutes, v2.ProfilingRoutes()...)
+	stdRoutes = append(stdRoutes, v2.LivenessCheckRoute(cfg.aliveCheckFunc))
+	stdRoutes = append(stdRoutes, v2.ReadyCheckRoute(cfg.readyCheckFunc))
+
+	for _, route := range stdRoutes {
+		handler := middleware.Chain(route.Handler(), middleware.NewRecovery())
+		mux.Handler(route.Method(), route.Path(), handler)
+		log.Debugf("added route %s", route)
+	}
+
+	// parse a list of HTTP numeric status codes that must be logged
+	statusCodeLoggerCfg, _ := os.LookupEnv("PATRON_HTTP_STATUS_ERROR_LOGGING")
+	statusCodeLogger, err := middleware.NewStatusCodeLoggerHandler(statusCodeLoggerCfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse status codes %s: %w", statusCodeLoggerCfg, err)
+	}
+
+	for _, route := range cfg.routes {
+		// add standard middlewares
+		middlewares := []middleware.Func{
+			middleware.NewRecovery(),
+			middleware.NewInjectObservability(),
+			middleware.NewLoggingTracing(route.Path(), statusCodeLogger),
+			middleware.NewRequestObserver(route.Method(), route.Path()),
+			middleware.NewCompression(cfg.deflateLevel),
+		}
+		// add router middlewares
+		middlewares = append(middlewares, cfg.middlewares...)
+		// add route middlewares
+		middlewares = append(middlewares, route.Middlewares()...)
+		// chain all middlewares to the handler
+		handler := middleware.Chain(route.Handler(), middlewares...)
+		mux.Handler(route.Method(), route.Path(), handler)
+		log.Debugf("added route %s with %d middlewares", route, len(middlewares))
+	}
+
+	return mux, nil
+}
+
+// Routes option for providing routes to the router.
+func Routes(routes ...*v2.Route) OptionFunc {
```

Comment:
I think this function is a bit limiting in real world, with bigger applications. I'll try to explain what I mean:

In the services where we use Patron, we often have multiple 'groups' of routes, that return a slice of Routes. I don't think there's a way to unpack multiple slices and pass them along in one go:

```Golang
loginController := login.NewController(someDepencency)
adminController := admin.NewController(someOtherDepencency)

router, err := phttprouter.New(phttprouter.Routes(
    // This is not possible:
    loginController.Routes..., 
    adminController.Routes...,
)) 
```

So you will always end up writing some intermediate state, something like this:

```Golang
loginController := login.NewController(someDepencency)
adminController := admin.NewController(someOtherDepencency)

routes := make([]*v2.Route, 0)
routes = append(routes, loginController.Routes()...)
routes = append(routes, adminController.Routes()...)

router, err := phttprouter.New(phttprouter.Routes(
    // You can only spread a single slice
    routes...,
)) 
```

I wonder how likely it is that a Patron service only has a single Controller, that returns a slice of routes. Maybe it makes more sense to have `httprouter.Routes` accept multiple slices, and 

Something like this:

```go
// Routes option for providing routes to the router.
 func Routes(routesCollections ...[]*v2.Route) OptionFunc {
        
 	// flatmap the many routes here
 	routes := make([]*v2.Route, 0)
 	for _, rr := range routesCollections {
 	        routes = append(routes, rr...)
        }

 	return func(cfg *Config) error {
 		if len(routes) == 0 {
 			return errors.New("routes are empty")
 		}
 		cfg.routes = routes
 		return nil
 	}
}
```

Reply:
Wouldn't a better option be to use the native [http.ServeMux](https://pkg.go.dev/net/http#example-ServeMux.Handle)? It should allow for some more route structuring.

Or we could go even more radical and expose `Handle` and `HandleFunc` methods of `srv` (see `createHTTPServer()`) directly, so that it is 100% compatible with anything and the kitchen sink.

Reply:
actually in your example you could do the following:

```
loginController := login.NewController(someDepencency)
adminController := admin.NewController(someOtherDepencency)

routes := append(loginController.Routes(), adminController.Routes()...)

router, err := phttprouter.New(phttprouter.Routes(
    // You can only spread a single slice
    routes...,
)) 
```
every subsequent slice of routes can appended to the routes.
Would this not do the trick?

Reply:
Sure, that would work, and if you have more than 2 collections of routes you can append them as well:

```go
loginController := login.NewController(someDepencency)
adminController := admin.NewController(someOtherDepencency)
otherController := other.NewController(otherDependency)
yetAnotherController := another.NewController(yetAnotherDependency)

routes := append(loginController.Routes(), adminController.Routes()...)
routes = append(routes, otherController.Routes()...)
routes = append(routes, yetAnotherController.Routes()...)

router, err := phttprouter.New(phttprouter.Routes(
    // You can only spread a single slice
    routes...,
)) 
```

But that was not my point :)

The `httprouter.Routes()` function takes one ore more `v2.Route` structs, which would allow you to use the method as follows:

```go
// With one route:
r1 := httprouter.Routes(aRoute)
// With multiple routes:
r2 := httprouter.Routes(routeOne, routeTwo, routeThree)
```

But from the examples I get the impression that you don't really create individual routes, but use the `v2.Routes` struct to collect multiple `v2.Route`s, which when using the `v2.Routes.Results` function will return `[]v2.*Route` slice.

So the return value of `v2.Routes.Result()` and argument of the argument of `httprouter.Routes` are different. This doesn't matter if it's only a single `[]v2.*Route` slice as you can unpack them, like you do the examples. 
But I think that in reality you will more likely have multiple `v2.Routes` objects, returning multiple `[]v2.*Route` slices.  So you end up writing an intermediate slice, like in the beginning of this message. 

Wouldn't it be better to move that logic into Patron, so it would be easier / less code to work with multiple `[]v2.*Route` slices?

For example, if you change the signature to this:

```go
func Routes(routesCollections ...[]*v2.Route) OptionFunc { ... }
```

The code would like this:
```go
loginController := login.NewController(someDepencency)
adminController := admin.NewController(someOtherDepencency)
otherController := other.NewController(otherDependency)
yetAnotherController := another.NewController(yetAnotherDependency)

router, err := phttprouter.New(phttprouter.Routes(
    loginController.Routes(), adminController.Routes(),  otherController.Routes(), yetAnotherController.Routes(),
)) 
```

Reply:
@peterklijn I did the change care to take a look? /cc @teivah @gmaz42 @tpaschalis

Reply:
Looks great!

Reply:
Personally, when, I read such a signature, I expect the variadic arguments and the slice to have a different meaning but here, it's just syntactic sugar from what I understand. In my opinion, the API looks a bit clumsy.

With variadic, we could still call the API in a single line using `append` if it's a requirement:

```go
router, err := New(append(loginController, adminController...)...)
```

Reply:
For me the code in original example is clearer/easier to read:
```go
loginController := login.NewController(someDepencency)
adminController := admin.NewController(someOtherDepencency)

routes := make([]*v2.Route, 0)
routes = append(routes, loginController.Routes()...)
routes = append(routes, adminController.Routes()...)

router, err := phttprouter.New(phttprouter.Routes(
    // You can only spread a single slice
    routes...,
))
```

Instead with the new syntax I expect people to fumble and not understand immediately the method signature (because of the variadic-of-slices nature).

Since we know that performance-wise it does not matter, I would opt for whichever version we think it's easier to read and understand.

Reply:
I prefer the original because I am convinced that `...[]*v2.Route` will be confusing for whoever tries to use this method :thinking: 

I am also concerned that if we are doing this change only because in Go it is not allowed, the syntax change could spread to other cases where we have `...SomeType` (technically all variadic signatures are affected by how the Go syntax works).

Reply:
Thanks for the feedback.
I will revert to the previous version of the code.
Thanks for the suggestion @peterklijn